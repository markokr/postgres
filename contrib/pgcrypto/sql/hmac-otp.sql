
-- HOTP (RFC4226):
--
-- Count    Hexadecimal    Decimal        HOTP
-- 0        4c93cf18       1284755224     755224
-- 1        41397eea       1094287082     287082
-- 2         82fef30        137359152     359152
-- 3        66ef7655       1726969429     969429
-- 4        61c5938a       1640338314     338314
-- 5        33c083d4        868254676     254676
-- 6        7256c032       1918287922     287922
-- 7         4e5b397         82162583     162583
-- 8        2823443f        673399871     399871
-- 9        2679dc69        645520489     520489

select c, hmac_otp('12345678901234567890', c, 'sha1', 10) as hotp10,
       hmac_otp('12345678901234567890', c, 'sha1', 6) as hotp6
    from generate_series(0, 9) c order by c;


-- TOTP (RFC6238):

-- +---------------+-----------------------+------------------+--------+--------+
-- |  Time(sec)    |   Time (UTC format)   | Value of T(Hex)  |  TOTP  | Mode   |
-- +---------------+-----------------------+------------------+--------+--------+
-- |  59           |  1970-01-01 00:00:59  | 0000000000000001 |94287082| SHA1   |
-- |  59           |  1970-01-01 00:00:59  | 0000000000000001 |46119246| SHA256 |
-- |  59           |  1970-01-01 00:00:59  | 0000000000000001 |90693936| SHA512 |
-- |  1111111109   |  2005-03-18 01:58:29  | 00000000023523EC |07081804| SHA1   |
-- |  1111111109   |  2005-03-18 01:58:29  | 00000000023523EC |68084774| SHA256 |
-- |  1111111109   |  2005-03-18 01:58:29  | 00000000023523EC |25091201| SHA512 |
-- |  1111111111   |  2005-03-18 01:58:31  | 00000000023523ED |14050471| SHA1   |
-- |  1111111111   |  2005-03-18 01:58:31  | 00000000023523ED |67062674| SHA256 |
-- |  1111111111   |  2005-03-18 01:58:31  | 00000000023523ED |99943326| SHA512 |
-- |  1234567890   |  2009-02-13 23:31:30  | 000000000273EF07 |89005924| SHA1   |
-- |  1234567890   |  2009-02-13 23:31:30  | 000000000273EF07 |91819424| SHA256 |
-- |  1234567890   |  2009-02-13 23:31:30  | 000000000273EF07 |93441116| SHA512 |
-- |  2000000000   |  2033-05-18 03:33:20  | 0000000003F940AA |69279037| SHA1   |
-- |  2000000000   |  2033-05-18 03:33:20  | 0000000003F940AA |90698825| SHA256 |
-- |  2000000000   |  2033-05-18 03:33:20  | 0000000003F940AA |38618901| SHA512 |
-- |  20000000000  |  2603-10-11 11:33:20  | 0000000027BC86AA |65353130| SHA1   |
-- |  20000000000  |  2603-10-11 11:33:20  | 0000000027BC86AA |77737706| SHA256 |
-- |  20000000000  |  2603-10-11 11:33:20  | 0000000027BC86AA |47863826| SHA512 |
-- +---------------+-----------------------+------------------+--------+--------+

create table algos ( algo text, seed bytea );
insert into algos values ('sha1', '12345678901234567890');
insert into algos values ('sha256', '12345678901234567890123456789012');
insert into algos values ('sha512', '1234567890123456789012345678901234567890123456789012345678901234');

create table times ( secs bigint );
insert into times values (59);
insert into times values (1111111109);
insert into times values (1111111111);
insert into times values (1234567890);
insert into times values (2000000000);
insert into times values (20000000000);

select t.secs, a.algo,
       hmac_otp(a.seed, t.secs / 30, a.algo, 6) as totp6,
       hmac_otp(a.seed, t.secs / 30, a.algo, 7) as totp7,
       hmac_otp(a.seed, t.secs / 30, a.algo, 8) as totp8
    from algos a, times t
    order by 1,2;

-- non-rfc values
select c, hmac_otp('12345678901234567890', c, 'sha224', 10) as hotp
    from generate_series(0, 9) c order by c;
select c, hmac_otp('12345678901234567890', c, 'sha384', 10) as hotp
    from generate_series(0, 9) c order by c;

