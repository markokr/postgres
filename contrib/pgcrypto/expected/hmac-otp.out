-- HOTP (RFC4226):
--
-- Count    Hexadecimal    Decimal        HOTP
-- 0        4c93cf18       1284755224     755224
-- 1        41397eea       1094287082     287082
-- 2         82fef30        137359152     359152
-- 3        66ef7655       1726969429     969429
-- 4        61c5938a       1640338314     338314
-- 5        33c083d4        868254676     254676
-- 6        7256c032       1918287922     287922
-- 7         4e5b397         82162583     162583
-- 8        2823443f        673399871     399871
-- 9        2679dc69        645520489     520489
select c, hmac_otp('12345678901234567890', c, 'sha1', 10) as hotp10,
       hmac_otp('12345678901234567890', c, 'sha1', 6) as hotp6
    from generate_series(0, 9) c order by c;
 c |   hotp10   | hotp6  
---+------------+--------
 0 | 1284755224 | 755224
 1 | 1094287082 | 287082
 2 | 0137359152 | 359152
 3 | 1726969429 | 969429
 4 | 1640338314 | 338314
 5 | 0868254676 | 254676
 6 | 1918287922 | 287922
 7 | 0082162583 | 162583
 8 | 0673399871 | 399871
 9 | 0645520489 | 520489
(10 rows)

-- TOTP (RFC6238):
-- +---------------+-----------------------+------------------+--------+--------+
-- |  Time(sec)    |   Time (UTC format)   | Value of T(Hex)  |  TOTP  | Mode   |
-- +---------------+-----------------------+------------------+--------+--------+
-- |  59           |  1970-01-01 00:00:59  | 0000000000000001 |94287082| SHA1   |
-- |  59           |  1970-01-01 00:00:59  | 0000000000000001 |46119246| SHA256 |
-- |  59           |  1970-01-01 00:00:59  | 0000000000000001 |90693936| SHA512 |
-- |  1111111109   |  2005-03-18 01:58:29  | 00000000023523EC |07081804| SHA1   |
-- |  1111111109   |  2005-03-18 01:58:29  | 00000000023523EC |68084774| SHA256 |
-- |  1111111109   |  2005-03-18 01:58:29  | 00000000023523EC |25091201| SHA512 |
-- |  1111111111   |  2005-03-18 01:58:31  | 00000000023523ED |14050471| SHA1   |
-- |  1111111111   |  2005-03-18 01:58:31  | 00000000023523ED |67062674| SHA256 |
-- |  1111111111   |  2005-03-18 01:58:31  | 00000000023523ED |99943326| SHA512 |
-- |  1234567890   |  2009-02-13 23:31:30  | 000000000273EF07 |89005924| SHA1   |
-- |  1234567890   |  2009-02-13 23:31:30  | 000000000273EF07 |91819424| SHA256 |
-- |  1234567890   |  2009-02-13 23:31:30  | 000000000273EF07 |93441116| SHA512 |
-- |  2000000000   |  2033-05-18 03:33:20  | 0000000003F940AA |69279037| SHA1   |
-- |  2000000000   |  2033-05-18 03:33:20  | 0000000003F940AA |90698825| SHA256 |
-- |  2000000000   |  2033-05-18 03:33:20  | 0000000003F940AA |38618901| SHA512 |
-- |  20000000000  |  2603-10-11 11:33:20  | 0000000027BC86AA |65353130| SHA1   |
-- |  20000000000  |  2603-10-11 11:33:20  | 0000000027BC86AA |77737706| SHA256 |
-- |  20000000000  |  2603-10-11 11:33:20  | 0000000027BC86AA |47863826| SHA512 |
-- +---------------+-----------------------+------------------+--------+--------+
create table algos ( algo text, seed bytea );
insert into algos values ('sha1', '12345678901234567890');
insert into algos values ('sha256', '12345678901234567890123456789012');
insert into algos values ('sha512', '1234567890123456789012345678901234567890123456789012345678901234');
create table times ( secs bigint );
insert into times values (59);
insert into times values (1111111109);
insert into times values (1111111111);
insert into times values (1234567890);
insert into times values (2000000000);
insert into times values (20000000000);
select t.secs, a.algo,
       hmac_otp(a.seed, t.secs / 30, a.algo, 6) as totp6,
       hmac_otp(a.seed, t.secs / 30, a.algo, 7) as totp7,
       hmac_otp(a.seed, t.secs / 30, a.algo, 8) as totp8
    from algos a, times t
    order by 1,2;
    secs     |  algo  | totp6  |  totp7  |  totp8   
-------------+--------+--------+---------+----------
          59 | sha1   | 287082 | 4287082 | 94287082
          59 | sha256 | 119246 | 6119246 | 46119246
          59 | sha512 | 693936 | 0693936 | 90693936
  1111111109 | sha1   | 081804 | 7081804 | 07081804
  1111111109 | sha256 | 084774 | 8084774 | 68084774
  1111111109 | sha512 | 091201 | 5091201 | 25091201
  1111111111 | sha1   | 050471 | 4050471 | 14050471
  1111111111 | sha256 | 062674 | 7062674 | 67062674
  1111111111 | sha512 | 943326 | 9943326 | 99943326
  1234567890 | sha1   | 005924 | 9005924 | 89005924
  1234567890 | sha256 | 819424 | 1819424 | 91819424
  1234567890 | sha512 | 441116 | 3441116 | 93441116
  2000000000 | sha1   | 279037 | 9279037 | 69279037
  2000000000 | sha256 | 698825 | 0698825 | 90698825
  2000000000 | sha512 | 618901 | 8618901 | 38618901
 20000000000 | sha1   | 353130 | 5353130 | 65353130
 20000000000 | sha256 | 737706 | 7737706 | 77737706
 20000000000 | sha512 | 863826 | 7863826 | 47863826
(18 rows)

-- non-rfc values
select c, hmac_otp('12345678901234567890', c, 'sha224', 10) as hotp
    from generate_series(0, 9) c order by c;
 c |    hotp    
---+------------
 0 | 0342893239
 1 | 0345812810
 2 | 0952291286
 3 | 1922303565
 4 | 0096943696
 5 | 0805550023
 6 | 1826794718
 7 | 0271708553
 8 | 0462957603
 9 | 1030641410
(10 rows)

select c, hmac_otp('12345678901234567890', c, 'sha384', 10) as hotp
    from generate_series(0, 9) c order by c;
 c |    hotp    
---+------------
 0 | 1860502125
 1 | 0646080675
 2 | 1187698930
 3 | 1223317355
 4 | 0837431422
 5 | 0218665174
 6 | 2121446854
 7 | 1737165020
 8 | 0727926248
 9 | 2146408529
(10 rows)

